### PR TITLE
Use `objdump` to determine whether a binary blob from a compressed archive should be handled by Carthage

### PR DESCRIPTION
**Pull request by blender**
Some frameworks don't have `Info.plists` and are being currently rejected as binary dependencies.

This PR addresses #2500 and also makes sure that .version files are written.

### CODE REVIEW

File: Source/CarthageKit/Project.swift [hunk 1]
```diff
@@ -1289,14 +1290,78 @@ internal func frameworksInDirectory(_ directoryURL: URL) -> SignalProducer<URL,
...
+					}
+					.filter {
+						$0 == MH_OBJECT // Carthage accepts static libraries
+						|| $0 == MH_BUNDLE // Bundles https://github.com/ResearchKit/ResearchKit/blob/1.3.0/ResearchKit/Info.plist#L15-L16
+						|| $0 == MH_DYLIB // or dynamic shared libraries
+					}
+					.single()?
+					.value
+
+					return fileType != nil
```
Comment (mdiep) [inline Source/CarthageKit/Project.swift]: I think this code ought to be split out into its own function.

File: Source/CarthageKit/VersionFile.swift [hunk 2]
```diff
@@ -49,7 +49,13 @@ struct VersionFile: Codable {
...
 	}
 
-	init(commitish: String, macOS: [CachedFramework]?, iOS: [CachedFramework]?, watchOS: [CachedFramework]?, tvOS: [CachedFramework]?) {
+	init(
+		commitish: String,
+		macOS: [CachedFramework]?,
+		iOS: [CachedFramework]?,
+		watchOS: [CachedFramework]?,
+		tvOS: [CachedFramework]?
+	) {
```
Comment (mdiep) [inline Source/CarthageKit/VersionFile.swift]: Are there any functional changes in this file? I'm having trouble spotting anything with the reformatting.

File: Source/CarthageKit/Project.swift (hunk 1 above)
Comment (mdiep) [reply Source/CarthageKit/Project.swift]: I'd want to see tests for this as well.

File: Source/CarthageKit/VersionFile.swift (hunk 2 above)
Comment (blender) [reply Source/CarthageKit/VersionFile.swift]: No there are no functional changes.

File: Source/CarthageKit/Project.swift (hunk 1 above)
Comment (blender) [reply Source/CarthageKit/Project.swift]: ok

File: Source/CarthageKit/Project.swift [hunk 3]
```diff
@@ -1289,14 +1290,78 @@ internal func frameworksInDirectory(_ directoryURL: URL) -> SignalProducer<URL,
...
+					.flatMap(.merge) { (output: String) -> SignalProducer<String, NoError> in
+						output.linesProducer
+					}.filter {
+						// Keep the lines with Mach-O magic number indefiers
+						$0.contains(String(format: "%x", MH_CIGAM))
+							|| $0.contains(String(format: "%x", MH_MAGIC))
+							|| $0.contains(String(format: "%x", MH_CIGAM_64))
+							|| $0.contains(String(format: "%x", MH_MAGIC_64))
+					}
+					.take(last: 1)
```
Comment (jdhealy) [inline Source/CarthageKit/Project.swift]: So, on the unstated assertion here (‘every `filetype` on a magic number containing line will be the same, so just grab any of them — as implemented: the last one’) — I said earlier that I’d post an implementation where a `reduce` makes this more explicit:

```swift
let acceptedFiletypesAsSpecifiedByMachHeader: Set<Int32> = [
	MH_OBJECT, // Carthage accepts static libraries
	MH_BUNDLE, // Bundles https://github.com/ResearchKit/ResearchKit/blob/1.3.0/ResearchKit/Info.plist#L15-L16
	MH_DYLIB, // or dynamic shared libraries
]

let øøø: Bool = task.launch(standardInput: nil)
	.ignoreTaskData()
	.map { String(data: $0, encoding: .utf8) ?? "" }
	.filter { !$0.isEmpty }
	.flatMap(.merge) { (output: String) -> SignalProducer<(String, String), NoError> in
		output.linesProducer.combinePrevious()
	}
	.filterMap {
		guard
			$0.0.hasPrefix("      magic cputype cpusubtype  caps    filetype ncmds"),
			[MH_CIGAM, MH_MAGIC, MH_CIGAM_64, MH_MAGIC_64]
				.lazy.map({ " " + String(format: "%x", $0) })
				.first(where: $0.1.hasPrefix) != nil
		else { return nil }

		let components = $0.1
			.components(separatedBy: CharacterSet.whitespaces)
			.filter { !$0.isEmpty }

		precondition(components.count == 8, "∆∆∆")

		return Int32(components[4])
	}
	.filter(acceptedFiletypesAsSpecifiedByMachHeader.contains)
	.reduce(into: Set<Int32>()) { $0.insert($1); return }
	.map { $0.count == 1 }
	.single()?
	.value ?? false
```

What do you think?

Comment (blender) [reply Source/CarthageKit/Project.swift]: I like the reduce but I think that this line with is maybe too "whitespace" sensitive?

`$0.0.hasPrefix("      magic cputype cpusubtype  caps    filetype ncmds"),`

Comment (jdhealy) [reply Source/CarthageKit/Project.swift]: Whitespace is fixed according to <https://llvm.org/viewvc/llvm-project/llvm/trunk/tools/llvm-objdump/MachODump.cpp?view=markup&pathrev=225383###see·line·2745> — IMO, we’d want a silent non-consideration if that output structure were to change (adding a field).

Comment (jdhealy) [reply Source/CarthageKit/Project.swift]: I guess we could strip the whitespace from both the output line and our conditional and compare, that would still catch an added field.

Comment (blender) [reply Source/CarthageKit/Project.swift]: I'm going to push a few changes according to @mdiep desire to have a separate function.

I'll try to put the reduce too

**Comment by jdhealy** [comment]
Great work tracking this down (I should have said upfront) — massive thanks 👏

**Comment by blender** [comment]
I have updated the PR:
- Introduced `MachHeader` strcture
- refactored out function to read mach headers in static method on `MachHeader`
- Added tests
- Remove checking for package type. It's superfluous give that the headers always work.

**Comment by blender** [comment]
Any opinion on the refactored version?

**Comment by mdiep** [comment]
The code looks fine to me, but I'm not totally confident whether this is a good idea. What do you think, @jdhealy?

File: Source/CarthageKit/Project.swift [hunk 4]
```diff
@@ -1,6 +1,7 @@
 // swiftlint:disable file_length
 
 import Foundation
+import MachO.loader
```
[the remark this answers is not in the document]
Comment (blender) [reply Source/CarthageKit/Project.swift]: No need. Good catch.

**Comment by blender** [comment]
The alternative that I have discussed with @jdhealy is to read the filetype by hand. I think it would be more error prone than using existing tools that know how to read Mach-O files correctly.

File: Source/CarthageKit/Project.swift [hunk 5]
```diff
@@ -1282,23 +1283,33 @@ internal func frameworksInDirectory(_ directoryURL: URL) -> SignalProducer<URL,
...
 			default:
-				return false
+				// In case no Info.plist exists check the Mach-O fileType
+				guard let executableURL = bundle?.executableURL else {
+					return false
+				}
+
+				return MachHeader.headers(forMachOFileAtUrl: executableURL)
+					.filter { MachHeader.carthageSupportedFileTypes.contains($0.fileType) }
+					.reduce(into: Set<UInt32>()) { $0.insert($1.fileType); return }
```
[the remark this answers is not in the document]
Comment (jdhealy) [reply Source/CarthageKit/Project.swift]: That `.insert` doesn’t return `Void` which necessitates either a `_ = ` prefix or a `return` suffix to match the type signature of the `into:` `reduce`.

File: Source/CarthageKit/MachHeader.swift [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+				guard
+					let cpuType = cpu_type_t(strippedComponents[1], radix: 10),
+					let cpuSubtype =  cpu_subtype_t(strippedComponents[2], radix: 10),
+					let fileType = UInt32(strippedComponents[4], radix: 10),
+					let ncmds = UInt32(strippedComponents[5], radix: 10),
+					let sizeofcmds = UInt32(strippedComponents[6], radix: 10),
+					let flags = UInt32(strippedComponents[7], radix: 16)
+					else {
+
```
Comment (jdhealy) [inline Source/CarthageKit/MachHeader.swift]: Nit: I’d prefer this without this empty line.

File: Source/CarthageKit/Project.swift [hunk 7]
```diff
@@ -1282,23 +1283,33 @@ internal func frameworksInDirectory(_ directoryURL: URL) -> SignalProducer<URL,
...
-
 			// Skip nested frameworks
 			let frameworksInURL = url.pathComponents.filter { pathComponent in
 				return (pathComponent as NSString).pathExtension == "framework"
 			}
 			return frameworksInURL.count == 1
 		}.filter { url in
-
-			let packageType: PackageType? = Bundle(url: url)?.packageType
+			// Try reading the Info.plist of the Framework
```
Comment (jdhealy) [inline Source/CarthageKit/Project.swift]: I’m confident in the structure of the CLI output from `objdump` rarely changing (and if it does, we should be able to catch via open-sourcedness before it’s widely distributed), but a simple Plist read is probably faster than a process execution. If you agree, could we leave this the default and add a comment with the reasoning?

```
// For reasons of speed and the fact that CLI-output structures can change,
// first try the safer method of reading the ‘Info.plist’ from the Framework’s bundle.
```

Comment (blender) [reply Source/CarthageKit/Project.swift]: Sure

File: Source/CarthageKit/MachHeader.swift [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+extension MachHeader {
+
+	/// Reads the Mach headers from a Mach-O file.
+	/// - Parameter url: The url of the Mach-O file
+	/// - Remark: Uses `objdump` to read the header and parse the output.
+	/// 	      The output is composed of one or more sets of lines like the following:
+	///
+	///			Mach header
+	///		          magic  cputype cpusubtype  caps    filetype ncmds sizeofcmds      flags
```
[the remark this answers is not in the document]
Comment (blender) [reply Source/CarthageKit/MachHeader.swift]: Something my new installation of Xcode is wrong and it's driving me crazy with spaces and tabs :D . Fixed.

File: Source/CarthageKit/MachHeader.swift (hunk 6 above)
Comment (blender) [reply Source/CarthageKit/MachHeader.swift]: Done

**Comment by blender** [comment]
I would like to rebase before merging to remove superfluous commits

**Comment by blender** [comment]
Rebased. All Done.

**Pull request merged**